### DISCUSSION TITLE
Warning: Invalid hook call. Hooks can only be called inside of the body of a function component

### DISCUSSION THREAD

**Discussion by JacobWeisenburger** [pmndrs/zustand#1712] [Q&A]
Anyone know what's causing this error?

```ts
import { createStore, useStore } from 'zustand'
import { render } from 'preact-render-to-string'

const fooStore = createStore( () => ( {
    foo: 'foo',
    bar: 42,
} ) )

function Component () {
    const foo = useStore( fooStore, state => state.foo )
    return <>foo: {foo}</>
}

console.log( render( <Component /> ) )
```

```
Warning: Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.
error: Uncaught TypeError: Cannot read properties of null (reading 'useRef')
```

**Comment by AVDiv** [comment]
Hope this might help:
https://legacy.reactjs.org/warnings/invalid-hook-call-warning.html

or else
what I feel is that the createStore() function should be replaced to create()

```
import { create, useStore } from 'zustand'
import { render } from 'preact-render-to-string'

const fooStore = create( () => ( {
    foo: 'foo',
    bar: 42,
} ) )
...
```

**Reply by JacobWeisenburger** [reply]
this didn't help. Same error in console.

**Comment by dbritto-dev** [comment]
@JacobWeisenburger which version of zustand are you using?

**Reply by JacobWeisenburger** [reply]
deno: 1.33.3 (release, x86_64-pc-windows-msvc)
v8: 11.4.183.2
typescript: 5.0.4
zustand: npm:zustand@4.3.8
preact-render-to-string: https://esm.sh/*preact-render-to-string@5.2.4

**Reply by dbritto-dev** [reply]
@JacobWeisenburger you need to alias `react` to `preact/compat` to make it work

**Reply by JacobWeisenburger** [reply]
can you send a full example, I'm not sure what you mean by alias `react`. Because there is no `react` in my example.

**Reply by dbritto-dev** [reply]
> can you send a full example, I'm not sure what you mean by alias `react`. Because there is no `react` in my example.

You are using `zustand` and `zustand` use `react`

So to being able to use `hooks` on `preact` you need to use `preact/compat`

Ref link: https://preactjs.com/guide/v10/getting-started/#aliasing-react-to-preact

**Reply by JacobWeisenburger** [reply]
> can you send a full example

**Reply by dbritto-dev** [reply]
> > can you send a full example

I have no idea how to do it on deno but on webpack you can use pkg json aliases

**Reply by JacobWeisenburger** [reply]
> I have no idea how to do it on deno

interesting that you give advice when you have no idea how to do it.

**Reply by dbritto-dev** [reply]
> > I have no idea how to do it on deno
> 
> interesting that you give advice when you have no idea how to do it.

That's because I know what's the issue, `zustand` still pointing to `react` but you are trying to use `preact` which only support `react` through `preact/compat` the only way to do it is by making `preact/compat` available as `react`. On webpack/node/snowpack is easy to do it on `deno` I suppose there's a way to achieve the same.

There's no science here, that's the expected behavior. You can fix your issue by importing `react` at top but then the result would be just and empty string. To make it fully work is necessary to use `preact/compat` instead of `react`.

How you can use `preact/compat` as `react` on deno is what you need to figure out.

**Reply by dbritto-dev** [reply]
BTW, I found this https://deno.com/manual@v1.32.2/basics/import_maps#overriding-imports
